core/src/test/scala/quasar/destination/gbq/GBQDestinationSpec.scala

```diff
@@ -0,0 +1,286 @@
+import java.util.UUID
+
+import org.http4s.argonaut.jsonEncoderOf
+import org.http4s.client.Client
+import org.http4s.{
+  AuthScheme, 
+  Credentials, 
+  EntityEncoder, 
+  Method, 
+  Request, 
+  Status,
+  Uri}
+import org.http4s.headers.Authorization
+import org.http4s.headers.`Content-Type`
+import org.http4s.client._
+
+import quasar.api.{Column, ColumnType}
+import quasar.api.destination.DestinationError.InitializationError
+import quasar.connector.destination.{Destination, ResultSink}
+import quasar.connector.render.RenderConfig
+import quasar.api.resource.{ResourceName, ResourcePath}
+import quasar.api.destination.{DestinationError, DestinationType}
+import quasar.connector.ResourceError
+import quasar.contrib.scalaz.MonadError_
+import quasar.EffectfulQSpec
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.ExecutionContext.Implicits.global
+
+import scala.{
+  Either,
+  List,
+  StringContext,
+}
+import scala.concurrent.duration._
+import scala.Predef.String
+import scala.util.Right
+import scala.util.Left
+
+import scalaz.{-\/,\/-}
+
+import slamdata.Predef.RuntimeException
+
+import shims._
+
+
+object GBQDestinationSpec extends EffectfulQSpec[IO] {
+  sequential
+
+  implicit val timer: Timer[IO] = IO.timer(ExecutionContext.global)
+
+  val testProject = "precog-ci-275718"
+  val testDataset = "dataset_" + UUID.randomUUID().toString.replace("-", "_").toString
+  val tableName = "table_" + UUID.randomUUID().toString.replace("-", "_").toString
+  val authCfgPath = Paths.get(getClass.getClassLoader.getResource("precog-ci-275718-e913743ebfeb.json").toURI)
+  val authCfgString = new String(Files.readAllBytes(authCfgPath), UTF_8)
+  val authCfgJson: Json = Parse.parse(authCfgString) match {
+    case Left(value) => Json.obj("malformed" := true)
+    case Right(value) => value
+  }
+  val gbqCfg = config(authCfgJson, testDataset)
+  val blockingPool = Executors.newFixedThreadPool(5)
+  val blocker = Blocker.liftExecutorService(blockingPool)
+  val httpClient: Client[IO] = JavaNetClientBuilder[IO](blocker).create
+
+  "csv link" should {
+    "reject empty paths with NotAResource" >>* {
+      csv(gbqCfg) { sink =>
+        val path = ResourcePath.root()
+        val req = sink.consume(path, NonEmptyList.one(Column("a", ColumnType.Boolean)), Stream.empty).compile.drain
+        MRE.attempt(req).map(_ must beLike {
+          case -\/(ResourceError.NotAResource(p2)) => p2 must_=== path
+        })
+      }
+    }
+
+    "successfully upload table" >>* {
+      csv(gbqCfg) { sink =>
+          val data = Stream("col1,col2\r\nstuff,true\r\n").through(text.utf8Encode)
+          val path = ResourcePath.root() / ResourceName(tableName) / ResourceName("bar.csv")
+          val req = sink.consume(
+            path,
+            NonEmptyList.fromList(List(Column("a", ColumnType.String), Column("b", ColumnType.Boolean))).get,
+            data).compile.drain
+
+          Timer[IO].sleep(5.seconds).flatMap { _ =>
```

Comment:
`5.seconds` probably should be a constant

Reply:
Is there a way to retry the operation, rather than sleep? Sleeping still means we're dependent on how long it takes the operation to complete, which occasionally could take much longer for reasons out of our control.

Reply:
And if it takes longer, it means that test would fail, and thus become a flaky test, which we want to avoid.